toolchain/parser/parse_tree.cpp

```diff
@@ -188,68 +188,84 @@ auto ParseTree::Print(llvm::raw_ostream& output, bool preorder) const -> void {
   output << "]\n";
 }
 
-auto ParseTree::Verify() const -> bool {
-  // Verify basic tree structure invariants.
-  llvm::SmallVector<ParseTree::Node, 16> ancestors;
-  for (Node n : llvm::reverse(postorder())) {
+auto ParseTree::Verify() const -> llvm::Optional<Error> {
+  llvm::SmallVector<ParseTree::Node> nodes;
+  // Traverse the tree in postorder.
+  for (Node n : postorder()) {
     const auto& n_impl = node_impls_[n.index()];
 
     if (n_impl.has_error && !has_errors_) {
-      llvm::errs()
-          << "Node #" << n.index()
-          << " has errors, but the tree is not marked as having any.\n";
-      return false;
+      return Error(llvm::formatv(
+          "Node #{0} has errors, but the tree is not marked as having any.",
+          n.index()));
     }
 
-    if (n_impl.subtree_size > 1) {
-      if (!ancestors.empty()) {
-        auto parent_n = ancestors.back();
-        const auto& parent_n_impl = node_impls_[parent_n.index()];
-        int end_index = n.index() - n_impl.subtree_size;
-        int parent_end_index = parent_n.index() - parent_n_impl.subtree_size;
-        if (parent_end_index > end_index) {
-          llvm::errs() << "Node #" << n.index() << " has a subtree size of "
-                       << n_impl.subtree_size
-                       << " which extends beyond its parent's (node #"
-                       << parent_n.index() << ") subtree (size "
-                       << parent_n_impl.subtree_size << ")\n";
-          return false;
+    int subtree_size = 1;
+    if (n_impl.kind.has_bracket()) {
+      while (true) {
+        if (nodes.empty()) {
+          return Error(
+              llvm::formatv("Node #{0} is a {1} with bracket {2}, but didn't "
+                            "find the bracket.",
+                            n.index(), n_impl.kind, n_impl.kind.bracket()));
+        }
+        auto child_impl = node_impls_[nodes.pop_back_val().index()];
+        subtree_size += child_impl.subtree_size;
+        if (n_impl.kind.bracket() == child_impl.kind) {
+          break;
         }
       }
-      // Has children, so we descend.
-      ancestors.push_back(n);
-      continue;
-    }
-
-    if (n_impl.subtree_size < 1) {
-      llvm::errs() << "Node #" << n.index()
-                   << " has an invalid subtree size of " << n_impl.subtree_size
-                   << "!\n";
-      return false;
-    }
-
-    // We're going to pop off some levels of the tree. Check each ancestor to
-    // make sure the offsets are correct.
-    int next_index = n.index() - 1;
-    while (!ancestors.empty()) {
-      ParseTree::Node parent_n = ancestors.back();
-      if ((parent_n.index() - node_impls_[parent_n.index()].subtree_size) !=
-          next_index) {
-        break;
+    } else if (n_impl.kind.child_count() == -2) {
+      // TODO: Allows -2 to mean "unsupported"; see parse_node_kind.def.
```

Comment:
Maybe just have a quick named constant in the kind type to make the meaning of -2 more obvious?

Reply:
Added TodoFixParseNode